numpy/core/src/umath/simd.inc.src

```diff
@@ -28,61 +28,57 @@
 #include <float.h>
 #include <string.h> /* for memcpy */
 
-/* Figure out the right abs function for pointer addresses */
-static NPY_INLINE npy_intp
-abs_intp(npy_intp x)
+static NPY_INLINE npy_uintp
+abs_ptrdiff(char *x, char *y)
 {
-#if (NPY_SIZEOF_INTP <= NPY_SIZEOF_INT)
-    return abs(x);
-#elif (NPY_SIZEOF_INTP <= NPY_SIZEOF_LONG)
-    return labs(x);
-#elif defined(_MSC_VER) && (_MSC_VER < 1600)
-    /* llabs is not available with Visual Studio 2008 */
-    return x > 0 ? x : -x;
-#else
-    return llabs(x);
-#endif
+    /* casting character pointers to npy_uintp is valid */
+    npy_uintp a = (npy_uintp)x;
+    npy_uintp b = (npy_uintp)y;
+
+    return (a > b) ? (a - b) : (b - a);
```

Comment:
Why even do the cast above?

Reply:
I wanted a check on the pointer type and thought it was clearer this way. The original had the cast in the signature, but this way seemed better. The integer cast is because taking the difference between pointers in different arrays is not well defined, although there may be an exception for `char *` pointer types.

Reply:
If course, if the difference between unrelated pointers breaks down, I expect NumPy as a whole to collapse into rubble.

Reply:
Strictly speaking, maybe should be `(npy_uintp)((void *)ptr)`. ANSII requires `char*` and `void*` to have the same representation, but I don't suppose that is necessarily the same for the result of casting them to integer.

Reply:
>the difference between pointers in different arrays is not well defined

I don't think the values of the integer-casted pointers is any more well-defined. Are the pointers unrelated here anyway? Supposedly section 6.3.2.3 of the C11 standard says:

> Except as previously specified, the result is implementation-defined, might not be correctly aligned, might not point to an entity of the referenced type, and might be a trap representation.